04-onion-routing.md

```diff
@@ -367,6 +381,239 @@ otherwise meets the amount criterion (eg. some other failure, or
+  - MUST use the decrypted fields to locate the next node
+- If `encrypted_data` contains a `next_blinding_override`:
+  - MUST use it as the next blinding point
+- Otherise:
+  - MUST use `E(i+1)` as the next blinding point
+- MUST forward the onion and communicate the next blinding point to the next node
+
+The final recipient:
+
+- MUST compute:
+  - `ss(r) = SHA256(k(r) * E(r))` (standard ECDH)
+  - `b(r) = HMAC256("blinded_node_id", ss(r)) * k(r)`
+  - `rho(r) = HMAC256("rho", ss(r))`
+- If an `encrypted_data` field is provided:
+  - MUST decrypt it using `rho(r)` and `E(r)`
+- MUST ignore the message if `E(r)` or `path_id` does not match the
+  blinded route it created
+
+#### Rationale
+
+Route blinding is a lightweight technique to provide recipient anonymity.
+It's more flexible than rendezvous because it lets senders choose what data
+they include in each node's onion payload and retry with updated values if
+the route fails.
+
+Use-cases where route blinding is useful include:
+
+- Recipient anonymity when receiving payments
+- Using unannounced channels in invoices without revealing them
+- Forcing a message to go through a specific set of intermediaries that can witness it
+- Providing anonymous reply paths for onion messages (offers, stuckless payments, etc)
+
+Each node in the blinded route needs to receive `E(i)` to be able to decrypt
+the onion and the `encrypted_data`. Protocols that use route blinding must
+specify how this value is propagated between nodes.
+
+When concatenating two blinded routes generated by different nodes, the
+last node of the first route needs to know the first `blinding` of the second
+route: the `next_blinding_override` field must be used to transmit this data.
+
+The final recipient must verify that the blinded route is used in the right
+context (e.g. for a specific payment). It can do so by storing `E(r)` and the
+context (e.g. a `payment_hash`), and verifying that they match when receiving
+the onion. Otherwise, to avoid additional storage cost, it can put some context
+information in the `path_id` field (e.g. a `payment_hash`) and verify that when
+receiving the onion.
+
+The `padding` field can be used to ensure that all `encrypted_data` have the
+same length. It's particularly useful when adding dummy hops at the end of a
+blinded route, to prevent the sender from figuring out which node is the final
+recipient.
+
+
+### Onion Message Payload Format
+
+Onion messages have an onion with an alternate `hop_payload`
+format: a `bigsize` followed by a `onionmsg_payload`.  Note that there
+is no legacy format, thus a `bigsize` of 0 means no payload.
+
+An intermediate node expects an `encrypted_data_tlv` which it can
+decrypt using the `blinding` which it is handed along with the onion
+message.
+
+Field numbers 64 and above are reserved for payloads for the final
+hop.
+
+1. `tlv_stream`: `onionmsg_payload`
+2. types:
+    1. type: 2 (`reply_path`)
+    2. data:
+        * [`point`:`first_node_id`]
```

Comment:
Discussed a while ago on [telegram](https://t.me/bolt12org/1687) -- do we want to get rid of this field and put the unblinded `first_node_id` in the blinded path's first hop? I believe offers encoding does this currently

Reply:
I thought @thomash-acinq pointed out that you can't do this without knowing the blinding secret :(  I think that may be correct...

Reply:
Ah! That checks out to me, thanks